<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ggheatmap: Tile-able heatmaps that play well with ggplots • ggheatmap</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="ggheatmap: Tile-able heatmaps that play well with ggplots">
<meta property="og:description" content="ggheatmap">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ggheatmap</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/ggheatmap.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ggheatmap_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>ggheatmap: Tile-able heatmaps that play well with ggplots</h1>
                        <h4 class="author">Clarice S. Groeneveld</h4>
            
            <h4 class="date">2021-04-26</h4>
      
      
      <div class="hidden name"><code>ggheatmap.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><code>ggheatmap</code> is built using <code>ggplot2</code> and the <code>patchwork</code> framework for aligning plots. It supports adding into the heatmap and all the powerful tweaks enabled by <code>theme</code>s in ggplot, and helps you align more information about your data to the original heatmap to make “complex heatmaps” with a more modern and <em>tweakable</em> framework.</p>
</div>
<div id="using-a-matrix-as-your-table" class="section level1">
<h1 class="hasAnchor">
<a href="#using-a-matrix-as-your-table" class="anchor"></a>Using a matrix as your table</h1>
<p>The simplest, yet less powerful, way to use <code>ggheatmap</code> is with a matrix as-is. Here, we demonstrate with a subset of a gene expression matrix for bladder cancer samples:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="co">#&gt; ── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──</span>
<span class="co">#&gt; ✓ ggplot2 3.3.3     ✓ purrr   0.3.4</span>
<span class="co">#&gt; ✓ tibble  3.1.0     ✓ dplyr   1.0.5</span>
<span class="co">#&gt; ✓ tidyr   1.1.3     ✓ stringr 1.4.0</span>
<span class="co">#&gt; ✓ readr   1.4.0     ✓ forcats 0.5.1</span>
<span class="co">#&gt; ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──</span>
<span class="co">#&gt; x dplyr::filter() masks stats::filter()</span>
<span class="co">#&gt; x dplyr::lag()    masks stats::lag()</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ggheatmap</span><span class="op">)</span>
<span class="co">#&gt; Registered S3 method overwritten by 'treeio':</span>
<span class="co">#&gt;   method     from</span>
<span class="co">#&gt;   root.phylo ape</span>

<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">tcgaBLCA_ex</span><span class="op">)</span>
<span class="va">gexp</span> <span class="op">&lt;-</span> <span class="va">tcgaBLCA_ex</span><span class="op">$</span><span class="va">gexp</span>

<span class="va">gghm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggheatmap.html">ggheatmap</a></span><span class="op">(</span><span class="va">gexp</span>,
          hm_colors <span class="op">=</span> <span class="st">'RdBu'</span>,
          hm_color_values <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html">rescale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">4</span>,<span class="op">-</span><span class="fl">2</span>,<span class="op">-</span><span class="fl">1</span>,<span class="op">-</span><span class="fl">0.5</span>,<span class="op">-</span><span class="fl">0.25</span>,<span class="fl">0</span>,<span class="fl">0.25</span>,<span class="fl">0.5</span>,<span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">6</span><span class="op">)</span><span class="op">)</span>,
          scale <span class="op">=</span> <span class="cn">TRUE</span>,
          center <span class="op">=</span> <span class="cn">TRUE</span>,
          show_dend_row <span class="op">=</span> <span class="cn">TRUE</span>,
          colors_title <span class="op">=</span> <span class="st">"Scaled expression (log2 UQ)"</span>,
          show_colnames <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">gghm</span></code></pre></div>
<p><img src="ggheatmap_files/figure-html/unnamed-chunk-1-1.png" width="576"> This method still will support extending with <code>align_to_hm</code> but would not support <code>add_tracks</code>. You can still modify this by using <code><a href="https://rdrr.io/r/base/Logic.html">&amp;</a></code> and adding <code>theme</code>, for example:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gghm</span> <span class="op">&amp;</span>
  <span class="fu">theme</span><span class="op">(</span>axis.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">6</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="ggheatmap_files/figure-html/unnamed-chunk-2-1.png" width="576"></p>
</div>
<div id="using-tables-with-more-variables" class="section level1">
<h1 class="hasAnchor">
<a href="#using-tables-with-more-variables" class="anchor"></a>Using tables with more variables</h1>
<p>If transpose our original table and extended it with new variables, we can unlock more features using <code>ggheatmap</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sample_annot</span> <span class="op">&lt;-</span> <span class="va">tcgaBLCA_ex</span><span class="op">$</span><span class="va">sample_annot</span> <span class="op">%&gt;%</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span>
<span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gexp</span><span class="op">)</span>
<span class="va">tcgaBLCA_tb</span> <span class="op">&lt;-</span> <span class="va">gexp</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">rownames_to_column</span><span class="op">(</span><span class="st">"sample"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">left_join</span><span class="op">(</span><span class="va">sample_annot</span>, by <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">group_by</span><span class="op">(</span><span class="va">consensusClass</span><span class="op">)</span>
<span class="va">tcgaBLCA_tb</span>
<span class="co">#&gt; # A tibble: 100 x 79</span>
<span class="co">#&gt; # Groups:   consensusClass [6]</span>
<span class="co">#&gt;    sample        FBP1 ACER2  PKHD1 CAPN5 S100P TMEM51 DHRS2 CYP4F22 SPINK1 ACSL5</span>
<span class="co">#&gt;    &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt;  1 TCGA-XF-A9S…  3.05 0.760 1.03    3.19  7.25   4.70 3.55   0.0458  8.07   5.22</span>
<span class="co">#&gt;  2 TCGA-BL-A13…  2.54 4.44  0.250   2.98  4.28   2.69 8.25   1.73    3.90   4.77</span>
<span class="co">#&gt;  3 TCGA-C4-A0F…  1.53 2.04  3.09    2.47  4.09   3.34 4.32   0.226   1.80   1.61</span>
<span class="co">#&gt;  4 TCGA-BT-A20…  5.98 5.05  1.22    4.92  7.03   4.97 8.03   5.10    8.85   6.83</span>
<span class="co">#&gt;  5 TCGA-GV-A40…  6.03 3.95  2.84    7.06  6.68   5.81 4.73   5.15    9.80   3.99</span>
<span class="co">#&gt;  6 TCGA-HQ-A2O…  5.56 5.36  3.59    5.41  7.05   5.54 6.58   1.12    7.46   6.39</span>
<span class="co">#&gt;  7 TCGA-XF-A8H…  8.32 5.76  3.46    6.90  8.26   5.50 9.91   7.05   11.2    6.65</span>
<span class="co">#&gt;  8 TCGA-DK-A2I…  7.42 4.78  1.18    5.45  7.82   4.89 8.41   5.38    8.42   7.50</span>
<span class="co">#&gt;  9 TCGA-UY-A78…  6.32 1.55  0.0317  1.64  7.23   5.51 8.22   1.82    0.474  4.05</span>
<span class="co">#&gt; 10 TCGA-BL-A13…  2.50 0.438 0.0181  3.88  1.16   4.88 0.696  0.451   0.204  5.02</span>
<span class="co">#&gt; # … with 90 more rows, and 68 more variables: ST3GAL5 &lt;dbl&gt;, TBX3 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   HPGD &lt;dbl&gt;, TGFBR3 &lt;dbl&gt;, FAM3B &lt;dbl&gt;, ATP8B1 &lt;dbl&gt;, RNF128 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   SNCG &lt;dbl&gt;, SLC44A3 &lt;dbl&gt;, GATA3 &lt;dbl&gt;, PPARG &lt;dbl&gt;, ICA1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   GGT6 &lt;dbl&gt;, RAB11A &lt;dbl&gt;, TRAK1 &lt;dbl&gt;, VSIG2 &lt;dbl&gt;, BCAS1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   RAB15 &lt;dbl&gt;, FAM174B &lt;dbl&gt;, SLC29A3 &lt;dbl&gt;, FOXA1 &lt;dbl&gt;, GOLT1A &lt;dbl&gt;,</span>
<span class="co">#&gt; #   PPFIBP2 &lt;dbl&gt;, DENND2D &lt;dbl&gt;, ACAA1 &lt;dbl&gt;, DNAJA4 &lt;dbl&gt;, HMGCS2 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   CYP2J2 &lt;dbl&gt;, VSNL1 &lt;dbl&gt;, KRT14 &lt;dbl&gt;, TGM1 &lt;dbl&gt;, SERPINB4 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   GSDMC &lt;dbl&gt;, KRT6A &lt;dbl&gt;, LGALS7 &lt;dbl&gt;, SFN &lt;dbl&gt;, SPRR2A &lt;dbl&gt;,</span>
<span class="co">#&gt; #   C12orf54 &lt;dbl&gt;, SPRR2D &lt;dbl&gt;, HOXD11 &lt;dbl&gt;, KRT6C &lt;dbl&gt;, KRT5 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   DSG3 &lt;dbl&gt;, KRT6B &lt;dbl&gt;, HOXD10 &lt;dbl&gt;, IL20RB &lt;dbl&gt;, RHCG &lt;dbl&gt;,</span>
<span class="co">#&gt; #   AHNAK2 &lt;dbl&gt;, SPRR2F &lt;dbl&gt;, FGFBP1 &lt;dbl&gt;, sex &lt;chr&gt;, age &lt;dbl&gt;,</span>
<span class="co">#&gt; #   stage &lt;chr&gt;, node &lt;chr&gt;, metastasis &lt;chr&gt;, papillary &lt;chr&gt;, squamous &lt;chr&gt;,</span>
<span class="co">#&gt; #   neuroendocrine &lt;chr&gt;, plasmacytoid &lt;chr&gt;, consensusClass &lt;fct&gt;,</span>
<span class="co">#&gt; #   cor_pval &lt;dbl&gt;, separationLevel &lt;dbl&gt;, LumP &lt;dbl&gt;, LumNS &lt;dbl&gt;, LumU &lt;dbl&gt;,</span>
<span class="co">#&gt; #   Stroma.rich &lt;dbl&gt;, Ba.Sq &lt;dbl&gt;, NE.like &lt;dbl&gt;</span></code></pre></div>
<p>One of the additional features we unlock is using <code>grouping</code> with <code>group_by</code> to make semi-supervised heatmaps. In this case, we need to say which column contains the IDs that will be the <em>columns</em> of the heatmap (in this case, <code>colv = 'sample'</code>), and what are the columns we want to plot as <em>rows</em> (<code>rowv = genes</code>). Other parameters here are graphical, for demonstration:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gr_gghm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggheatmap.html">ggheatmap</a></span><span class="op">(</span><span class="va">tcgaBLCA_tb</span>,
    colv <span class="op">=</span> <span class="st">"sample"</span>,
    rowv <span class="op">=</span> <span class="va">genes</span>,
    hm_colors <span class="op">=</span> <span class="st">'RdBu'</span>,
    hm_color_values <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html">rescale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">4</span>,<span class="op">-</span><span class="fl">2</span>,<span class="op">-</span><span class="fl">1</span>,<span class="op">-</span><span class="fl">0.5</span>,<span class="op">-</span><span class="fl">0.25</span>,<span class="fl">0</span>,<span class="fl">0.25</span>,<span class="fl">0.5</span>,<span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">6</span><span class="op">)</span><span class="op">)</span>,
    scale <span class="op">=</span> <span class="cn">TRUE</span>,
    center <span class="op">=</span> <span class="cn">TRUE</span>,
    show_dend_row <span class="op">=</span> <span class="cn">FALSE</span>,
    show_colnames <span class="op">=</span> <span class="cn">FALSE</span>,
    show_rownames <span class="op">=</span> <span class="cn">FALSE</span>,
    group_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>`Ba/Sq` <span class="op">=</span> <span class="st">"#fe4a49"</span>, LumNS <span class="op">=</span> <span class="st">"#32837d"</span>, LumP <span class="op">=</span> <span class="st">"#06d6a0"</span>, LumU <span class="op">=</span> <span class="st">"#009fb7"</span>,
                     `Stroma-rich` <span class="op">=</span> <span class="st">"#f9c80e"</span>, `NE-like` <span class="op">=</span> <span class="st">"#7d5ba6"</span><span class="op">)</span>,
    colors_title <span class="op">=</span> <span class="st">"Scaled expression (log2 UQ)"</span><span class="op">)</span>

<span class="va">gr_gghm</span> <span class="op">+</span>
  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">'collect'</span><span class="op">)</span></code></pre></div>
<p><img src="ggheatmap_files/figure-html/unnamed-chunk-4-1.png" width="672"> Here, we can see that the heatmap is clustered in a semi-supervised manner (by <code>consensusClass</code>). It also enables <code>add_tracks</code>:</p>
<div id="adding-tracks" class="section level2">
<h2 class="hasAnchor">
<a href="#adding-tracks" class="anchor"></a>Adding tracks</h2>
<p>You can <code>add_tracks</code> for variables that were in the original table fed to <code>ggheatmap</code>. You can see what variables are available using <code>gghmData</code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/gghmData.html">gghmData</a></span><span class="op">(</span><span class="va">gr_gghm</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt;  [1] "observations"    "FBP1"            "ACER2"           "PKHD1"          </span>
<span class="co">#&gt;  [5] "CAPN5"           "S100P"           "TMEM51"          "DHRS2"          </span>
<span class="co">#&gt;  [9] "CYP4F22"         "SPINK1"          "ACSL5"           "ST3GAL5"        </span>
<span class="co">#&gt; [13] "TBX3"            "HPGD"            "TGFBR3"          "FAM3B"          </span>
<span class="co">#&gt; [17] "ATP8B1"          "RNF128"          "SNCG"            "SLC44A3"        </span>
<span class="co">#&gt; [21] "GATA3"           "PPARG"           "ICA1"            "GGT6"           </span>
<span class="co">#&gt; [25] "RAB11A"          "TRAK1"           "VSIG2"           "BCAS1"          </span>
<span class="co">#&gt; [29] "RAB15"           "FAM174B"         "SLC29A3"         "FOXA1"          </span>
<span class="co">#&gt; [33] "GOLT1A"          "PPFIBP2"         "DENND2D"         "ACAA1"          </span>
<span class="co">#&gt; [37] "DNAJA4"          "HMGCS2"          "CYP2J2"          "VSNL1"          </span>
<span class="co">#&gt; [41] "KRT14"           "TGM1"            "SERPINB4"        "GSDMC"          </span>
<span class="co">#&gt; [45] "KRT6A"           "LGALS7"          "SFN"             "SPRR2A"         </span>
<span class="co">#&gt; [49] "C12orf54"        "SPRR2D"          "HOXD11"          "KRT6C"          </span>
<span class="co">#&gt; [53] "KRT5"            "DSG3"            "KRT6B"           "HOXD10"         </span>
<span class="co">#&gt; [57] "IL20RB"          "RHCG"            "AHNAK2"          "SPRR2F"         </span>
<span class="co">#&gt; [61] "FGFBP1"          "sex"             "age"             "stage"          </span>
<span class="co">#&gt; [65] "node"            "metastasis"      "papillary"       "squamous"       </span>
<span class="co">#&gt; [69] "neuroendocrine"  "plasmacytoid"    "consensusClass"  "cor_pval"       </span>
<span class="co">#&gt; [73] "separationLevel" "LumP"            "LumNS"           "LumU"           </span>
<span class="co">#&gt; [77] "Stroma.rich"     "Ba.Sq"           "NE.like"</span></code></pre></div>
<p><code>observations</code> will always be your ID variable, which is a factor ordered in the same way as the heatmap, to ease making new plots that will be perfectly aligned. Here, we’ll add some clinical tracks:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gr_gghm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_tracks.html">add_tracks</a></span><span class="op">(</span><span class="va">gr_gghm</span>,
           track_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"stage"</span>, <span class="st">"node"</span>, <span class="st">"metastasis"</span><span class="op">)</span>,
           track_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>stage <span class="op">=</span> <span class="st">'Greys'</span>, node <span class="op">=</span> <span class="st">'Oranges'</span>, metastasis <span class="op">=</span> <span class="st">'Reds'</span><span class="op">)</span>,
           track_prop <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>
<span class="co">#&gt; Adding missing grouping variables: `consensusClass`</span>
<span class="va">gr_gghm</span> <span class="op">+</span>
  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">'collect'</span><span class="op">)</span></code></pre></div>
<p><img src="ggheatmap_files/figure-html/unnamed-chunk-6-1.png" width="672"></p>
</div>
<div id="aligning-new-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#aligning-new-plots" class="anchor"></a>Aligning new plots</h2>
<p>Here we’ll make two different plots: one to align with the samples (columns) and another to align with the genes (rows). We can get the data from gghmData, which is a good idea to facilitate further plotting because it will ensure your observations will be in the correct order. Then, here, we make a line-plot with the correlations of each sample to the centroid of the 6 consensus bladder cancer subtypes. Note the use of <code>theme_quant</code>, one of our suggested themes that look nice with <code>ggheatmap</code>s, and that we switch the y-axis to the right side to make the end-product look better (though everything will work without this):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tcgaBLCA_tb2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gghmData.html">gghmData</a></span><span class="op">(</span><span class="va">gr_gghm</span><span class="op">)</span>
<span class="va">plt_corlines</span> <span class="op">&lt;-</span> <span class="va">tcgaBLCA_tb2</span> <span class="op">%&gt;%</span>
  <span class="fu">ungroup</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">select</span><span class="op">(</span><span class="va">observations</span>, <span class="va">LumP</span><span class="op">:</span><span class="va">NE.like</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">pivot_longer</span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">observations</span>, names_to <span class="op">=</span> <span class="st">"subtype"</span>, values_to <span class="op">=</span> <span class="st">"cor"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">observations</span>, <span class="va">cor</span>, color <span class="op">=</span> <span class="va">subtype</span>, group <span class="op">=</span> <span class="va">subtype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">scale_y_continuous</span><span class="op">(</span>position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>`Ba.Sq` <span class="op">=</span> <span class="st">"#fe4a49"</span>, LumNS <span class="op">=</span> <span class="st">"#32837d"</span>, LumP <span class="op">=</span> <span class="st">"#06d6a0"</span>, LumU <span class="op">=</span> <span class="st">"#009fb7"</span>,
                                `Stroma.rich` <span class="op">=</span> <span class="st">"#f9c80e"</span>, `NE.like` <span class="op">=</span> <span class="st">"#7d5ba6"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">guides</span><span class="op">(</span>color <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"Correlation\n to centroid"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/theme_sparse.html">theme_quant</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">plt_corlines</span></code></pre></div>
<p><img src="ggheatmap_files/figure-html/unnamed-chunk-7-1.png" width="480"> The second plot should align to the rows. We can’t just use the <code>gghmData</code> because it doesn’t contain the information to order our rows, but we can get this from <code>gghm_rowLevels</code>. Here, we plot which signature each gene in our example data belongs to:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plt_row_annot</span> <span class="op">&lt;-</span> <span class="va">tcgaBLCA_ex</span><span class="op">$</span><span class="va">gene_annot</span> <span class="op">%&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>gene_symbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">gene_symbol</span>, levels <span class="op">=</span> <span class="fu"><a href="../reference/gghmData.html">gghm_rowLevels</a></span><span class="op">(</span><span class="va">gr_gghm</span><span class="op">)</span><span class="op">)</span>,
         group <span class="op">=</span> <span class="st">'signature'</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">gene_symbol</span>, <span class="va">group</span>, fill <span class="op">=</span> <span class="va">signature</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_tile</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">coord_flip</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/theme_sparse.html">theme_sparse2</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">plt_row_annot</span></code></pre></div>
<p><img src="ggheatmap_files/figure-html/unnamed-chunk-8-1.png" width="288"></p>
<p>Finally, we can use <code>align_to_hm</code> to add these plots to the original hm with all panels properly aligned. Note the use of <code>legend_action = 'collect'</code> in the final call, that will unite all the legends in a nice way:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gghm_complete</span> <span class="op">&lt;-</span> <span class="va">gr_gghm</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/align_to_hm.html">align_to_hm</a></span><span class="op">(</span><span class="va">plt_corlines</span>, newplt_size_prop <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/align_to_hm.html">align_to_hm</a></span><span class="op">(</span><span class="va">plt_row_annot</span>, pos <span class="op">=</span> <span class="st">"left"</span>, newplt_size_prop <span class="op">=</span> <span class="fl">0.08</span>, 
              legend_action <span class="op">=</span> <span class="st">"collect"</span>, tag_level <span class="op">=</span> <span class="st">'keep'</span><span class="op">)</span>
<span class="va">gghm_complete</span></code></pre></div>
<p><img src="ggheatmap_files/figure-html/unnamed-chunk-9-1.png" width="816"></p>
<p>Note that you can still use the <code>patchwork</code> <code><a href="https://rdrr.io/r/base/Logic.html">&amp;</a></code> to make global changes to all plots:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gghm_complete</span> <span class="op">&lt;-</span> <span class="va">gghm_complete</span> <span class="op">&amp;</span>
  <span class="fu">theme</span><span class="op">(</span>legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">7</span><span class="op">)</span>,
        legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span><span class="op">)</span>
<span class="va">gghm_complete</span></code></pre></div>
<p><img src="ggheatmap_files/figure-html/unnamed-chunk-10-1.png" width="768"></p>
</div>
</div>
<div id="creating-panels" class="section level1">
<h1 class="hasAnchor">
<a href="#creating-panels" class="anchor"></a>Creating panels</h1>
<p>You can now make panels with your complete heatmap with <code>patchwork</code>, <code>cowplot</code> or other aligning packages. For example:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plt_subtype_count</span> <span class="op">&lt;-</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">sample_annot</span>, <span class="fu">aes</span><span class="op">(</span><span class="va">consensusClass</span>, fill <span class="op">=</span> <span class="va">consensusClass</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_bar</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>`Ba/Sq` <span class="op">=</span> <span class="st">"#fe4a49"</span>, LumNS <span class="op">=</span> <span class="st">"#32837d"</span>, 
                               LumP <span class="op">=</span> <span class="st">"#06d6a0"</span>, LumU <span class="op">=</span> <span class="st">"#009fb7"</span>,
                                `Stroma-rich` <span class="op">=</span> <span class="st">"#f9c80e"</span>, 
                               `NE-like` <span class="op">=</span> <span class="st">"#7d5ba6"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">'Number of samples'</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">guides</span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="../reference/theme_sparse.html">theme_quant</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme</span><span class="op">(</span>axis.ticks.x <span class="op">=</span> <span class="fu">element_line</span><span class="op">(</span>color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>,
        axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>color <span class="op">=</span> <span class="st">"black"</span>, angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span>

<span class="va">plt_subtype_count</span></code></pre></div>
<p><img src="ggheatmap_files/figure-html/unnamed-chunk-11-1.png" width="384"></p>
<p>Then, we can just use standard <code>patchwork</code> to align the <code>ggheatmap</code> and our new plots (as they won’t be aligned with the heatmap part of the plot, but with the entire plot):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span>
<span class="va">new_col</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">plt_subtype_count</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_spacer.html">plot_spacer</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.3</span>,<span class="fl">0.7</span><span class="op">)</span><span class="op">)</span>

<span class="op">(</span><span class="va">new_col</span> <span class="op">|</span> <span class="va">gghm_complete</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.4</span>,<span class="fl">0.6</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="ggheatmap_files/figure-html/unnamed-chunk-12-1.png" width="1152"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Clarice Groeneveld.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
